internal/server/node/property_value.go

```diff
@@ -108,12 +113,57 @@ func GetPropertyValuesHelper(
 	arcOut bool,
 ) (map[string][]*model.Node, error) {
 	rowList := bigtable.BuildPropertyValuesKey(dcids, prop, arcOut)
-	// Current branch cache is targeted on new stats (without addition of schema etc),
-	// so only use base cache data for property value.
-	//
-	// TODO(shifucun): perform a systematic check on current cache data and see
-	// if this is still true.
-	return readPropertyValues(ctx, store, rowList)
+	baseDataList, _, err := bigtable.Read(
+		ctx,
+		store.BtGroup,
+		rowList,
+		func(dcid string, jsonRaw []byte) (interface{}, error) {
+			var propVals model.PropValueCache
+			err := json.Unmarshal(jsonRaw, &propVals)
+			if err != nil {
+				return nil, err
+			}
+			return propVals.Nodes, nil
+		},
+		nil,
+		true, /* readBranch */
+	)
+	if err != nil {
+		return nil, err
+	}
+	result := map[string][]*model.Node{}
+	visited := map[string]map[string]struct{}{}
+	// Loop the import groups. They are ordered by the preferences. So only add
+	// a node if it is not seen yet.
+	for _, baseData := range baseDataList {
+		for dcid, data := range baseData {
+			if _, ok := result[dcid]; !ok {
+				result[dcid] = []*model.Node{}
+			}
+			if _, ok := visited[dcid]; !ok {
+				visited[dcid] = map[string]struct{}{}
+			}
+			if data != nil {
+				nodes := data.([]*model.Node)
+				for _, n := range nodes {
+					// Check if a duplicate node has been added to the result.
+					// Duplication is based on either the DCID or the value.
+					if n.Dcid != "" {
+						if _, ok := visited[dcid][n.Dcid]; !ok {
+							result[dcid] = append(result[dcid], n)
+							visited[dcid][n.Dcid] = struct{}{}
+						}
+					} else if n.Value != "" {
```

Comment:
To confirm, are we combining values?  I thought we would pick all values of a property from the preferred cache..

Reply:
Yes, the loop starts from the preferred cache. New node is only added when it is not in visited.